File: in_Python/0020 Valid Parentheses.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stack = []
        for x in list(s):
            if x in ['[','{','(']:
                stack.append(x)
            else:
                if x == ']' and len(stack) != 0 and stack[-1] == '[':
                    stack.pop()
                elif x == ')' and len(stack) != 0 and stack[-1] == '(':
                    stack.pop()
                elif x == '}' and len(stack) != 0 and stack[-1] == '{':
                    stack.pop()
                else:
                    return False
        return len(stack) == 0
        

File: in_Java/0020 Valid Parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> st = new Stack<>();
        
        for(char c : s.toCharArray()){
            if(c == '(' || c == '[' || c == '{') st.push(c);
            else{
                if(!st.isEmpty()){
                    char t = st.peek();
                    if(c == ')' && t == '(') st.pop();
                    else if(c == ']' && t == '[') st.pop();
                    else if(c == '}' && t == '{') st.pop();
                    else return false;
                }else{
                    return false;
                }            
            }
        }
        return st.isEmpty();
    }
}